desktopexporter/app/routes/main-view.tsx

```diff
@@ -27,50 +47,103 @@ export default function MainView() {
           isFullWidth={isFullWidth}
           toggleSidebarWidth={setFullWidth.toggle}
           traceSummaries={[]}
+          numNewTraces={0}
         />
         <EmptyStateView />
       </Flex>
     );
   }
 
-  let sidebarSummaries: TraceSummaryWithUIData[] =
-    getTraceSummariesWithUIData(traceSummaries);
   return (
     <Flex height="100vh">
       <Sidebar
         isFullWidth={isFullWidth}
         toggleSidebarWidth={setFullWidth.toggle}
-        traceSummaries={sidebarSummaries}
+        traceSummaries={sidebarData.summaries}
+        numNewTraces={sidebarData.numNewTraces}
       />
       <Outlet />
     </Flex>
   );
 }
 
-function getTraceSummariesWithUIData(
+function initSidebarData(traceSummaries: TraceSummary[]): SidebarData {
+  return {
+    summaries: traceSummaries.map((traceSummary) =>
+      generateTraceSummaryWithUIData(traceSummary),
+    ),
+    numNewTraces: 0,
+  };
+}
+
+function updateSidebarData(
+  sidebarData: SidebarData,
   traceSummaries: TraceSummary[],
-): TraceSummaryWithUIData[] {
-  return traceSummaries.map((traceSummary) => {
-    if (traceSummary.hasRootSpan) {
-      let duration = getDurationNs(
-        traceSummary.rootStartTime,
-        traceSummary.rootEndTime,
-      );
-
-      let durationString = getDurationString(duration);
-      return {
-        hasRootSpan: true,
-        rootServiceName: traceSummary.rootServiceName,
-        rootName: traceSummary.rootName,
-        rootDurationString: durationString,
-        spanCount: traceSummary.spanCount,
-        traceID: traceSummary.traceID,
-      };
+): SidebarData {
+  
+  let mergedData: SidebarData = {
+    numNewTraces: 0,
+    summaries: [...sidebarData.summaries],
+  };
+
+  // Check for new and stale traces
+  for (let summary of traceSummaries) {
+    let traceID = summary.traceID;
+    let sidebarSummaryIndex = mergedData.summaries.findIndex(
+      (s) => s.traceID === traceID,
+    );
+
+    if (sidebarSummaryIndex === -1) {
+      // If the traceID of the new summary has no match in the sidebar
+      // increment the number of new traces.
+      mergedData.numNewTraces++;
+    } else if (
+      summary.spanCount > mergedData.summaries[sidebarSummaryIndex].spanCount
+    ) {
+      // If the number of spans in an existing trace is greater than what's displayed in the sidebar
+      // generate a whole new summary with ui data
+      mergedData.summaries[sidebarSummaryIndex] =
+        generateTraceSummaryWithUIData(summary);
     }
+  }
+
+  // Check for deleted/expired traces
+  for (let i = 0; i < mergedData.summaries.length; i++) {
```

Comment:
This is fine, but FYI if you wanted to use for-of here and also get the index, you could use `.entries()`

```typescript
for (let [i, summary] of mergedData.summaries.entries()) {
  // ...
}
```

Reply:
Cool! TIL